awx/main/analytics/collectors.py

```diff
@@ -78,7 +80,51 @@ def events_slicing(key, since, until):
     lower = since or last_gather or horizon
     if not since and last_entries.get(key):
         lower = horizon
-    pk_values = models.JobEvent.objects.filter(created__gte=lower, created__lte=until).aggregate(Min('pk'), Max('pk'))
+    partition_epoch = get_event_partition_epoch()
+
+    def get_partitions():
+        """
+        Get all event partitions. Exclude _unpartitioned_main_jobevents partition if it exists.
+        """
+        with connection.cursor() as cursor:
+            cursor.execute(f"select to_regclass('_unpartitioned_main_jobevents') is not null")
+            old_events_table_exists = cursor.fetchone()[0]
+
+            query = f"SELECT inhrelid::regclass::text AS child FROM pg_catalog.pg_inherits WHERE inhparent = 'public.main_jobevent'::regclass"
+            if old_events_table_exists:
+                query += f" AND inhrelid != 'public._unpartitioned_main_jobevents'::regclass"
+
+            cursor.execute(query)
+            partitions_from_db = [r[0] for r in cursor.fetchall()]
+
+        return partitions_from_db
+
+    def partition_within_horizon(partition, horizon):
+        dt_str = partition[14:]
+        dt = datetime.datetime.strptime(dt_str, '%Y%m%d_%H').replace(tzinfo=pytz.UTC)
+        return dt > horizon
+
+    # TODO: Make default once unpartitioned tables are no longer supported
+    # if we are collecting job events starting at the horizon
+    # and all job events are stored in partitions,
+    # slice by partition
+    if lower == horizon and lower > partition_epoch:
```

Comment:
I don't agree with limiting this optimization to the case where the lower bound is exactly at the 4-week limit.

Reply:
agreed, rewriting this 👍